cluster/src/main/java/io/atomix/cluster/impl/BroadcastMemberLocationProvider.java

```diff
@@ -0,0 +1,114 @@
+import io.atomix.cluster.messaging.BroadcastService;
+import io.atomix.utils.event.AbstractListenerManager;
+import io.atomix.utils.net.Address;
+import io.atomix.utils.serializer.Namespace;
+import io.atomix.utils.serializer.Namespaces;
+import io.atomix.utils.serializer.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.atomix.utils.concurrent.Threads.namedThreads;
+
+/**
+ * Member location provider implementation that uses the broadcast service.
+ */
+public class BroadcastMemberLocationProvider
+    extends AbstractListenerManager<MemberLocationEvent, MemberLocationEventListener>
+    implements MemberLocationProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(BroadcastMemberLocationProvider.class);
+  private static final Serializer SERIALIZER = Serializer.using(Namespace.builder()
+      .register(Namespaces.BASIC)
+      .nextId(Namespaces.BEGIN_USER_CUSTOM_ID)
+      .register(new AddressSerializer(), Address.class)
+      .build());
+
+  private final BroadcastService broadcastService;
+  private final boolean enabled;
+  private final ScheduledExecutorService broadcastScheduler = Executors.newSingleThreadScheduledExecutor(
+      namedThreads("atomix-cluster-broadcast", LOGGER));
+  private final Duration broadcastInterval = Duration.ofSeconds(5);
+  private volatile ScheduledFuture<?> broadcastFuture;
+  private final Consumer<byte[]> broadcastListener = this::handleBroadcastMessage;
+  private final Set<Address> locations = Sets.newConcurrentHashSet();
+
+  public BroadcastMemberLocationProvider(BroadcastService broadcastService, boolean enabled) {
+    this.broadcastService = checkNotNull(broadcastService);
+    this.enabled = enabled;
+  }
+
+  private void handleBroadcastMessage(byte[] message) {
+    Address address = SERIALIZER.decode(message);
+    if (locations.add(address)) {
+      post(new MemberLocationEvent(MemberLocationEvent.Type.JOIN, address));
+    }
+  }
+
+  private void broadcastLocation(Address address) {
+    broadcastService.broadcast(SERIALIZER.encode(address));
+  }
+
+  @Override
+  public CompletableFuture<Void> join(Address address) {
+    if (!enabled) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    locations.add(address);
+    broadcastService.addListener(broadcastListener);
+    broadcastFuture = broadcastScheduler.scheduleWithFixedDelay(
+        () -> broadcastLocation(address),
+        broadcastInterval.toMillis(),
+        broadcastInterval.toMillis(),
+        TimeUnit.MILLISECONDS);
+    broadcastLocation(address);
```

Comment:
do we need to broadcast the location twice? you are already broadcasting the location on line 91.

Reply:
I see now, you register the broadcast and this line is for triggering the broadcast now.